pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,267 @@
+    if (quotaConfig.isMaxQueriesPerSecondValid()) {
+      overallRate = Double.parseDouble(quotaConfig.getMaxQueriesPerSecond());
+    } else {
+      LOGGER.error("Failed to init qps quota: error when parsing qps quota: {} for table: {}",
+          quotaConfig.getMaxQueriesPerSecond(), tableName);
+      return;
+    }
+
+    double perBrokerRate =  overallRate / onlineCount;
+    QueryQuotaConfig queryQuotaConfig = new QueryQuotaConfig(RateLimiter.create(perBrokerRate), new HitCounter(TIME_RANGE_IN_SECOND));
+    _rateLimiterMap.put(tableName, queryQuotaConfig);
+    LOGGER.info("Rate limiter for table: {} has been initialized. Overall rate: {}. Per-broker rate: {}. Number of online broker instances: {}",
+        tableName, overallRate, perBrokerRate, onlineCount);
+  }
+
+  /**
+   * Acquire a token from rate limiter based on the table name.
+   * @param tableName original table name which could be raw.
+   * @return true if there is no query quota specified for the table or a token can be acquired, otherwise return false.
+   * */
+  public boolean acquire(String tableName, String offlineTableName, String realtimeTableName) {
+    LOGGER.info("Trying to acquire token for table: {}", tableName);
+
+    QueryQuotaConfig offlineTableQueryQuotaConfig = null;
+    QueryQuotaConfig realtimeTableQueryQuotaConfig = null;
+    if (offlineTableName != null) {
+      offlineTableQueryQuotaConfig = _rateLimiterMap.get(offlineTableName);
+    }
+    if (realtimeTableName != null) {
+      realtimeTableQueryQuotaConfig = _rateLimiterMap.get(realtimeTableName);
+    }
+
+    if (offlineTableQueryQuotaConfig == null && realtimeTableQueryQuotaConfig == null) {
+      LOGGER.info("No qps quota is specified for table: {}", tableName);
+      return true;
+    }
+
+    // Try to acquire token for offline table if there's a qps quota.
+    if (offlineTableQueryQuotaConfig != null && !tryAcquireToken(tableName, offlineTableQueryQuotaConfig)) {
+      return false;
+    }
+
+    // Try to acquire token for offline table if there's a qps quota.
+    return realtimeTableQueryQuotaConfig == null || tryAcquireToken(tableName, realtimeTableQueryQuotaConfig);
+  }
+
+  /**
+   * Try to acquire token from rate limiter. Emit the utilization of the qps quota if broker metric isn't null.
+   * @param tableName origin table name, which could be raw.
+   * @param queryQuotaConfig query quota config for type-specific table.
+   * @return true if there's no qps quota for that table, or a token is acquired successfully.
+   * */
+  private boolean tryAcquireToken(String tableName, QueryQuotaConfig queryQuotaConfig) {
+    // Use hit counter to count the number of hits.
+    queryQuotaConfig.getHitCounter().hit();
+
+    RateLimiter rateLimiter = queryQuotaConfig.getRateLimiter();
+    double perBrokerRate = rateLimiter.getRate();
+    if (!rateLimiter.tryAcquire()) {
+      LOGGER.error("Quota is exceeded for table: {}. Per-broker rate: {}", tableName, perBrokerRate);
```

Comment:
this is not an ERROR log. should be info.

Also, bump a metric when quota is exceeded

Reply:
Yup, a metric called "QUERY_QUOTA_EXCEEDED" will be emitted in `BrokerRequestHandler`.